usr/src/uts/common/fs/zfs/dsl_scan.c

```diff
@@ -1891,3 +2409,1016 @@ dsl_scan_restarting(dsl_scan_t *scn, dmu_tx_t *tx)
 	return (scn->scn_restart_txg != 0 &&
 	    scn->scn_restart_txg <= tx->tx_txg);
 }
+
+/*
+ * Grand theory statement on scan queue sorting
+ *
+ * Scanning is implemented by recursively traversing all indirection levels
+ * in an object and reading all blocks referenced from said objects. This
+ * results in us approximately traversing the object from lowest logical
+ * offset to the highest. Naturally, if we were simply read all blocks in
+ * this order, we would require that the blocks be also physically arranged
+ * in sort of a linear fashion on the vdevs. However, this is frequently
+ * not the case on pools. So we instead stick the I/Os into a reordering
+ * queue and issue them out of logical order and in a way that most benefits
+ * physical disks (LBA-order).
+ *
+ * This sorting algorithm is subject to limitations. We can't do this with
+ * blocks that are non-leaf, because the scanner itself depends on these
+ * being available ASAP for further metadata traversal. So we exclude any
+ * block that is bp_level > 0. Fortunately, this usually represents only
+ * around 1% of our data volume, so no great loss.
+ *
+ * As a further limitation, we cannot sort blocks which have more than
+ * one DVA present (copies > 1), because there's no sensible way to sort
+ * these (how do you sort a queue based on multiple contradictory
+ * criteria?). So we exclude those as well. Again, these are very rarely
+ * used for leaf blocks, usually only on metadata.
```

Comment:
I think the change would be pretty small, and worthwhile because it would speed up the traversal (block discovery).  You would need to create a qzio_t (or whatever we're calling them now) for each DVA.